Port CheckPileup

magicDGS opened the pull request:
Because I'm implementing a program in GATK that comes for a pipeline using `samtools mpileup` and post-processing the file, I need a way to check if the internal `ReadPileup` generated by the tool is indeed the same as the one obtained for samtools.

Thus, I ported the `CheckPileup` tool from GATK3 in this PR.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+ *
+ * <p>At every locus in the input set, compares the pileup data (reference base, aligned base from
+ * each overlapping read, and quality score) generated internally by GATK to a reference pileup data generated
+ * by Samtools. Note that the pileup program has been replaced in Samtools by mpileup, which produces a slightly
```
Does your tool read mpileup format?  Samtools no longer offers the pileup command.  It looks like mpileup produces something very similar to the 6 column format here, but I'm not sure what the differences if any are.

magicDGS replied:
mpileup and pileup are exactly the same, except the consensus format (not supported). I will remove all code related with "consensus pileup".

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+ * different output format by default.
+ * </p>
+ *
+ * <h3>Format</h3>
```
No need to include a description of the format here.  At most have a link to the pileup format definition.

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+    FeatureInput<SAMPileupFeature> pileup;
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "An output file created by the walker. Will overwrite contents if file exists")
+    private PrintStream out;
```
Huh, I didn't realize PrintStream had a string constructor.  This saves a few lines, but it's probably better use a `File` here and manually instantiate the stream in order to be able to throw a clearer `UserException` when an invalid path is given.  Right now it will result in something like 
`A USER ERROR has occurred: Invalid command line: Argument  output has a bad value: /some/invalid/path. Problem constructing PrintStream from the string '/some/invalid/path'.` 

Which isn't that bad, but would be clearer as a `CouldNotCreateOutputFile`  Not a big deal either way.  There's a long neglected plan to make generating and cleaning up output streams cleaner, but it hasn't happened yet.

lbergelson replied:
It might be nice to make this output file optional, and have the diff go to stdout if it's not specified.  Fine like this though.

magicDGS replied:
Done both

magicDGS replied:
When I ported `Pileup` (#1836), I was using also a `PrintStream` (it was thw way GATK3 implement it). Should I open a ticket for change it, @lbergelson? I think your point is right, could clarify user errors...

lbergelson replied:
No harm in adding a ticket.  Low priority, but it might help push us into coming up with a better solution where the engine can handle instantiating output streams and cleaning up after them.

magicDGS replied:
Created issue

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+     * to the next position.
+     */
+    @Argument(fullName="continue_after_error",doc="Continue after encountering an error",optional=true)
+    public boolean continueAfterAnError = false;
```
it seems strange that all the arguments have different levels of visibility.

magicDGS replied:
All `public` now.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+        return true;
+    }
+
+    public long nLoci = 0;
```
these should probably be private

magicDGS replied:
Done (also nBases)

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    /**
+     * Extracts the true pileup data from the given rodSAMPileup.  Note that this implementation
```
Update these references to "rods"

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+        if( pileupArg == null)
+            return null;
+
+        if( pileupArg.hasPointGenotype() ) {
```
this logic is strange, getPointGenotype just returns  `isPointGenotype() ? this : null` and so does getIndelGenotypes

and it looks like `hasPointGenotype` is defined as `!hasIndelGenotype`

I think this whole block can just be `return pileupArg == null ? null : return pileupArg`  

I don't see how that exception could ever be thrown...  Maybe I'm just missing something dumb though.

magicDGS replied:
Simplified, because I'm going to get rid of the consensus format.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+    public String pileupDiff(final ReadPileup a, final SAMPileupFeature b, boolean orderDependent)
+    {
+        if ( a.size() != b.size() ) {
+            return "Sizes not equal";
```
It would be nice to include the sizes so you can see how unequal they are, as well as maybe including the position.

magicDGS replied:
Included.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+            return "Sizes not equal";
+        }
+        if (IntervalUtils.compareLocatables(a.getLocation(), b, getReferenceDictionary()) != 0 ) {
+            return "Locations not equal";
```
same for the rest of these, probably want more information in the error message so people can see the difference.  You're probably the primary consumer of this though, so do what you think would be useful to you.

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+        return x;
+    }
+
+    public String pileupDiff(final ReadPileup a, final SAMPileupFeature b, boolean orderDependent)
```
since the only usage of this method is hardcoded with `orderDependent = true` I would either remove the parameter and simplify the code, or expose it as a command line parameter.

magicDGS replied:
Added TODO, I will decide later if it is worthy to maintain.

lbergelson commented:
As part of this, could you port `SAMPileupCodecUnitTest`.  It's shamefully incomplete and only tests `canDecode`, but that's better than nothing.  It would be good to have some unit tests for the codec itself on various pileup lines the different parsing cases.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,226 @@
+            return "Bases not equal";
+        }
+
+        String aQuals = maybeSorted(new String(a.getBaseQuals()), ! orderDependent );
```
This has the strange property that if these were sorted the quals can be sorted differently than the bases.  It seems like they should be sorted together because each qual goes with a specific base.

magicDGS replied:
Added TODO regarding this issue (I will try to fix using `PileupElement` lists)

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,343 @@
+
+ * <h3>Format</h3>
+ * <p>There are two versions of the original pileup format: the current 6-column format produced by Samtools, and the old
+ * 10/13-column "consensus" format which could be obtained by using the -c argument, now deprecated. </p>
```
do we need to support the deprecated format?  It seems like it would be better not to unless you really need it for something.

magicDGS replied:
It's not going to be supported, I deleted the documentation.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,343 @@
+ * <p>The Pileup format was first used by Tony Cox and Zemin Ning at
+ *     the Sanger Institute. It describes the base-pair information at each chromosomal position. This format
+ *     facilitates SNP/indel calling and brief alignment viewing by eye. Note that the pileup program has been replaced
+ *     in Samtools by mpileup, which produces a slightly different output format by default.
```
Do we handle single sample mpileups?  We definitely should since it's the standard now.

magicDGS replied:
Noted that the codec is only for single-sample mpileup

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,343 @@
+ * </p>
+ *
+ * <h3>Caveat</h3>
+ * <p>Handling of indels is questionable at the moment. Proceed with care.</p>
```
this makes me uncomfortable...

magicDGS replied:
I removed the whole reference to the format, and I'll implement this from scratch.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,343 @@
+    private static final String emptyStr = ""; // we will use this for "reference" allele in insertions
+
+    // codec file extension
+    protected static final String FILE_EXT = "samp";
```
I think we may also want to accept `.pileup` since that's what the wikipedia page recommends as the most popular file extension for pileups.

magicDGS replied:
I removed "samp" (breaking compatibility with GATK3) and add the recomended ones ("pileup" and "mpileup")

lbergelson replied:
👍 to that decision

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,343 @@
+                    feature.setEnd(feature.getStart());
+                }
+                break;
+            case consensusSNPTokenCount: // pileup called a SNP or a reference base
```
unless you need them I would just remove the parsing for deprecated pileups and greatly simplify this class

magicDGS replied:
Removed.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupFeature.java:
```diff
@@ -0,0 +1,289 @@
+ * @version 0.1
+ */
+public class SAMPileupFeature implements Feature {
+    public enum VariantType { NONE, SNP, INSERTION, DELETION, INDEL };
```
If it's ok to drop the deprecated pileup formats than much of this class can also be dropped.

magicDGS replied:
Removed.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupFeature.java:
```diff
@@ -0,0 +1,289 @@
+     *
+     * @return
+     */
+    public SAMPileupFeature getIndelGenotype() {
```
a lot of these are for some outdated GenotypeList interface that aren't needed anymore and don't really do anything

magicDGS replied:
Removed.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupFeature.java:
```diff
@@ -0,0 +1,289 @@
+     * @return
+     */
+    public boolean hasIndelGenotype() {
+        return isIndelGenotype();
```
is indel genotype and has indel genotype are redundant

magicDGS replied:
Removed both (not needed anymore).

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupFeature.java:
```diff
@@ -0,0 +1,289 @@
+     *
+     * @return
+     */
+    public boolean hasPointGenotype() {
```
same with hasPointGenotype

magicDGS replied:
Removed both.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileupIntegrationTest.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Test(enabled = false)
+    public void testConsensusPileup() throws IOException {
+        // TODO
```
If you're not including deprecated pileups than you remove this case

magicDGS replied:
Removed

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileupIntegrationTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * This test runs on a basic pileup obtained with samtools (version 1.1) and options -B --min-BQ 0
```
would you mind using a more recent samtools to regenerate the test files and use mpileup?

magicDGS replied:
It should generate exactly the same result, but I will do it asap when I finish the implementation.

lbergelson commented:
@magicDGS Review complete for now.  Looks good but I have some nitpicks.  I think they're almost all due to it being ancient gatk3 code that no one has updated in a long time.  I'd recommend dropping the deprecated formats and only supporting mpilup single sample format which should allow for massive simplification of both the Codec and the Feature.  

We need some unit tests for the codec itself since it has a bunch of different potential error cases, and we should have some integration tests for the tool that show it correctly failing on cases with errors.

magicDGS commented:
Thanks, @lbergelson. I was also thinkinhg that this code is mostly deprecated, but I wanted to ported as is for the first pass review. I just need to support the new mpileup version (unique sample, because if not it is more difficult), because the consensus one is deprecated.

I will update the codec and add some tests for it. In addition, I was thinking to create a list of `PileupElement` inside the feature to make easier to compare the internal pileup, but with "reads" of one base-pair.

On the other hand, I will improve the walker itself. I will tell you when I finished with the changes.

lbergelson commented:
@magicDGS I appreciate the minimal diff initial import, makes it much easier to see what happened.  It's going to be much cleaner once you knock out all of that deprecated code.

vdauwera commented:
FYI the last person who touched this in GATK3 was probably me -- I had to fix a bug in handing both old & new pileup format a while ago. I remember cringing a lot but my mission was to get it to not crash -- cleaning it up further was way out of scope. 

I say that purely defensively in case anybody does git blame and sees my name pop up ;)

magicDGS commented:
Ready for second pass review, @lbergelson. Now the implementation is much more simple than the previous one, and I added unit tests for the codec.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -0,0 +1,164 @@
+    public void onTraversalStart() {
+        try {
+            out = (outFile == null) ? System.out : new PrintStream(outFile);
+        } catch (FileNotFoundException e) {
```
Sorry for this extra boiler plate.  We should improve the output stream handling in the future  part of #121 )

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+    // including also space as delimiter
+    private static final Pattern SPLIT_PATTERN = Pattern.compile("\\t|( +)");
+
+    // although the minimum number of fields is 6 (see specifications), if the coverage is 0 sometimes bases and qualities does not appear
```
thanks for this helpful explanation

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+     */
+    @Override
+    public boolean canDecode(final String path) {
+        return SAM_PILEUP_FILE_EXTENSIONS.stream().anyMatch(ext -> path.endsWith("."+ext));
```
could you make this case insensitive?

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+        final String[] tokens = SPLIT_PATTERN.split(line.trim(), -1);
+        // check the number of fields
+        if(tokens.length < MINIMUM_FIELDS || tokens.length > MAXIMUM_FIELDS) {
+            throw new CodecLineParsingException("The SAM pileup line didn't have the expected number of columns. " +
```
Could you include the actual text of the line in this comment as well as the number of columns and the expected range?

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+                    "Note that this codec is only valid for single-sample pileups.");
+        }
+        // starting parsing
+        final int pos = parseInteger(tokens[1], "position");
```
Would you mind extracting the contig out here too for consitency?

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+
+    @VisibleForTesting
+    protected List<SAMPileupElement> parseBasesAndQuals(final String bases, final String qualities, final byte ref) {
+        // TODO: probably a index out of bounds exception will be thrown
```
Is this still relevant?

magicDGS replied:
It is because I forgot to add the try-catch block. If the qualities have less length than the bases because the pileup is malformed, the code won't reach the check of the lengths at the end of the method. I've just added the try-catch

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodec.java:
```diff
@@ -0,0 +1,166 @@
+                    final String rest = bases.substring(i+1);
+                    final Matcher match = INDEL_REGEXP.matcher(rest);
+                    if(! match.matches()) {
+                        throw new CodecLineParsingException("The SAM pileup line have an indel marker (+/-) without lenght");
```
typo:  "The SAM pileup line *has* an indel marker (+/-) without *length*"

magicDGS replied:
Corrected

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileupIntegrationTest.java:
```diff
@@ -0,0 +1,38 @@
+            " -pileup " +  TEST_OUTPUT_DIRECTORY + "reads_data_source_test1.samtools.samp" +
+            " -O %s", Arrays.asList(emptyTemp.toString()));
+
+        testSpec.executeTest("testBasicPileup", this);
```
Could you add an additional test that shows it correctly catching a difference?

magicDGS replied:
Done, with BAQ qualities differences.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileupIntegrationTest.java:
```diff
@@ -0,0 +1,38 @@
+    private static final String TEST_OUTPUT_DIRECTORY = publicTestDir + "org/broadinstitute/hellbender/tools/walkers/qc/pileup/";
+
+    /**
+     * This test runs on a basic pileup obtained with samtools (version 1.1) and options -B --min-BQ 0
```
is this still accurate?

magicDGS replied:
No, now it with samtools version 1.3.1. Changed

lbergelson commented on src/test/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupCodecUnitTest.java:
```diff
@@ -0,0 +1,112 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class SAMPileupCodecUnitTest extends BaseTest {
```
Could you add a test that uses `decodeLoc`?

magicDGS replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/sampileup/SAMPileupFeature.java:
```diff
@@ -0,0 +1,136 @@
+     * Returns formatted pileup string for the genomic location as
+     * "location: reference_base observed_base_pile observed_qual_pile"
+     */
+    public String getPileupString() {
```
Could you add a test for this method?  It's super simple but it's important that it doesn't change unexpectedly and people tend to be cavalier about toString sorts of things.

magicDGS replied:
It is weird that `ReadPileup` and `SAMPileupFeature`, which are similar, have a different behavior regarding this method. I changed now the contract and added a test for it (and solve a bug that I detected in getQualsAsString).

magicDGS replied:
Actually they are representing the same kind of feature, but `ReadPileup` backed the reads I don't know if it is worthy, because `SAMPileupFeature` will be used only for this tool. From my side, it is just a intermediate step to be sure that the new GATK framework will provide me an easy way of create internally exactly the same pileups as samtools. And it will be more important when #589 provides a way of filter in a similar way as samtools and and #1886 expose the option to do BAQ. Anyway, in my next commit the methods will be normalized (names) and we can discuss about the interface.

lbergelson commented:
@magicDGS You really made it much clearner.  Huge improvement over the old code.  Thank you! .  

Back to you with a few typos and request for a few more tests.

magicDGS commented:
Thanks for the review, @lbergelson. Back to you with addressed comments and added tests.

I don't know if it is good for the clarity of the framework to create an interface for `Pileup`s (for `ReadPileup` and `SAMPileupFeature`), but let me know if you would like that I include that in this PR.

lbergelson commented:
@magicDGS It sounds like a common interface isn't necessary. 

Do you think this codec would be useful in htsjdk?  Would other people want to use it?  We could consider moving it there in the future.

lbergelson commented:
This looks good to me.  👍

lbergelson commented:
@magicDGS Thanks for the contribution!

magicDGS commented:
Thanks for reviewing, @lbergelson. I think that the codec could be useful in htsjdk, because I know several people that works with the samtools-pileup format, and it will be nice to be able to use it in the htsjdk API. In addition, it could be useful for include new behaviour and refine the implementation (like retrieving reads from the format).
